File: app/src/main/java/com/denizd/substitutionplan/CardAdapter.java
package com.denizd.substitutionplan;

import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

public class CardAdapter extends RecyclerView.Adapter<CardAdapter.CardViewHolder> {
    private List<Subst> mSubst;
    private static int colour;
    private static String colorCheck;

    public static class CardViewHolder extends RecyclerView.ViewHolder {
        public ImageView mImageView;
        public TextView mGroup, mDate, mTime, mCourse, mRoom, mAdditional;

        public CardViewHolder(@NonNull View itemView) {
            super(itemView);
            mImageView = itemView.findViewById(R.id.iconView);
            mGroup = itemView.findViewById(R.id.group);
            mDate = itemView.findViewById(R.id.date);
            mTime = itemView.findViewById(R.id.time);
            mCourse = itemView.findViewById(R.id.course);
            mRoom = itemView.findViewById(R.id.room);
            mAdditional = itemView.findViewById(R.id.additional);
        }
    }

    public CardAdapter(ArrayList<Subst> subst) {
        mSubst = subst;
    }

    @NonNull
    @Override
    public CardViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.course_card, parent, false);
        CardViewHolder cvh = new CardViewHolder(v);
        return cvh;
    }

    public Subst getSubstAt(int i) {
        return mSubst.get(i);
    }

    @Override
    public void onBindViewHolder(@NonNull CardViewHolder holder, int position) {
        Subst currentItem = mSubst.get(position);
        final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(holder.mImageView.getContext());


        holder.mImageView.setImageResource(currentItem.getIcon());
        holder.mGroup.setText(currentItem.getGroup());
        holder.mDate.setText(currentItem.getDate());
        holder.mTime.setText(currentItem.getTime());
        holder.mCourse.setText(currentItem.getCourse());
        holder.mRoom.setText(currentItem.getRoom());
        holder.mAdditional.setText(currentItem.getAdditional());

        colour = 0;
        try {
            colorCheck = holder.mCourse.getText().toString().toLowerCase().substring(0, 3);
            switch (colorCheck) {
                case "deu": case "dep": case "daz": case "fda": {
                    colour = prefs.getInt("colGerman", 0);
                    break;
                }
                case "mat": case "map": {
                    colour = prefs.getInt("colMaths", 0);
                    break;
                }
                case "eng": case "enp": case "ena": {
                    colour = prefs.getInt("colEnglish", 0);
                    break;
                }
                case "spo": case "spp": case "spth": {
                    colour = prefs.getInt("colPhysEd", 0);
                    break;
                }
                case "pol": case "pop": {
                    colour = prefs.getInt("colPolitics", 0);
                    break;
                }
                case "dar": case "dap": {
                    colour = prefs.getInt("colTheatre", 0);
                    break;
                }
                case "phy": case "php": {
                    colour = prefs.getInt("colPhysics", 0);
                    break;
                }
                case "bio": case "bip": {
                    colour = prefs.getInt("colBiology", 0);
                    break;
                }
                case "che": case "chp": {
                    colour = prefs.getInt("colChemistry", 0);
                    break;
                }
                case "phi": case "psp": {
                    colour = prefs.getInt("colPhilosophy", 0);
                    break;
                }
                case "laa": case "laf": case "lat": {
                    colour = prefs.getInt("colLatin", 0);
                    break;
                }
                case "spa": case "spf": {
                    colour = prefs.getInt("colSpanish", 0);
                    break;
                }
                case "fra": case "frf": case "frz": {
                    colour = prefs.getInt("colFrench", 0);
                    break;
                }
                case "inf": {
                    colour = prefs.getInt("colCompsci", 0);
                    break;
                }
                case "ges": {
                    colour = prefs.getInt("colHistory", 0);
                    break;
                }
                case "rel": {
                    colour = prefs.getInt("colReligion", 0);
                    break;
                }
                case "geg": case "wuk": {
                    colour = prefs.getInt("colGeography", 0);
                    break;
                }
                case "kun": {
                    colour = prefs.getInt("colArts", 0);
                    break;
                }
                case "mus": {
                    colour = prefs.getInt("colMusic", 0);
                    break;
                }
                case "tue": {
                    colour = prefs.getInt("colTurkish", 0);
                    break;
                }
                case "chi": {
                    colour = prefs.getInt("colChinese", 0);
                    break;
                }
                case "gll": {
                    colour = prefs.getInt("colGLL", 0);
                    break;
                }
                case "wat": {
                    colour = prefs.getInt("colWAT", 0);
                    break;
                }
                case "för": {
                    colour = prefs.getInt("colForder", 0);
                    break;
                }
                case "met": {
                    colour = prefs.getInt("colWP", 0);
                    break;
                }
                default: {
                    colour = 0;
                    break;
                }
            }
        } catch (StringIndexOutOfBoundsException e) {
            try {
                String colorCheck2 = holder.mCourse.getText().toString().toLowerCase().substring(0, 2);
                switch (colorCheck2) {
                    case "nw": {
                        colour = prefs.getInt("colBiology", 0);
                        break;
                    }
                    case "wp": {
                        colour = prefs.getInt("colWP", 0);
                        break;
                    }
                    default: {
                        colour = 0;
                        break;
                    }
                }
            } catch (StringIndexOutOfBoundsException e1) {
                colour = 0;
            }
        } finally {
            if (colour == 0) {
                if (prefs.getInt("themeInt", 0) == 1) {
                    colour = R.color.lightgrey;
                } else {
                    colour = R.color.lessdark;
                }
            }
            holder.mImageView.getDrawable().setTint(ContextCompat.getColor(holder.mImageView.getContext(), colour));
            holder.mCourse.setTextColor(ContextCompat.getColor(holder.mCourse.getContext(), colour));
        }
    }

    @Override
    public int getItemCount() {
        return mSubst.size();
    }

    public void setSubst(List<Subst> subst) {
        mSubst = subst;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/denizd/substitutionplan/CardAdapter.kt
package com.denizd.substitutionplan

import android.content.ActivityNotFoundException
import android.net.Uri
import android.preference.PreferenceManager
import android.text.SpannableString
import android.text.style.StrikethroughSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.card.MaterialCardView
import java.util.*

class CardAdapter(private var mSubst: List<Subst>) : RecyclerView.Adapter<CardAdapter.CardViewHolder>() {

    private var colour = 0
    private var colourString = ""
    private var colorCheck = ""

    class CardViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {

        var mImageView: ImageView = itemView.findViewById(R.id.iconView)
        var mGroup: TextView = itemView.findViewById(R.id.group)
        var mDate: TextView = itemView.findViewById(R.id.date)
        var mTime: TextView = itemView.findViewById(R.id.time)
        var mCourse: TextView = itemView.findViewById(R.id.course)
        var mRoom: TextView = itemView.findViewById(R.id.room)
        var mAdditional: TextView = itemView.findViewById(R.id.additional)
        var mCard: MaterialCardView = itemView.findViewById(R.id.planCard)
        init { view.setOnClickListener(this) }
        override fun onClick(v: View?) {
            if (mDate.text.toString().length > 7 && mDate.text.toString().substring(3, 7) == "http") {
                try {
                    CustomTabsIntent.Builder().build().launchUrl(mCard.context,
                            Uri.parse(mDate.text.toString().substring(3)))
                } catch (e: ActivityNotFoundException) {
                    Toast.makeText(mCard.context, mCard.context.getString(R.string.chromeCompatibleNotFound), Toast.LENGTH_LONG).show()
                }
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CardViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.course_card, parent, false)
        return CardViewHolder(v)
    }

    override fun onBindViewHolder(holder: CardViewHolder, position: Int) {
        val currentItem = mSubst[position]
        val prefs = PreferenceManager.getDefaultSharedPreferences(holder.mImageView.context)
        var psa = false
        val strings = arrayOf(SpannableString(currentItem.group), SpannableString(currentItem.time),
                SpannableString(currentItem.course), SpannableString(currentItem.room))
        for (item in strings) {
            val qmark = item.indexOf("?")
            if (qmark != -1) {
                item.setSpan(StrikethroughSpan(), 0, qmark, 0)
            }
        }
        with (currentItem.additional.toLowerCase(Locale.ROOT)) {
            if (contains("eigenverantwortliches arbeiten") || contains("entfall")) {
                strings[2].setSpan(StrikethroughSpan(), 0, strings[2].length, 0)
                strings[3].setSpan(StrikethroughSpan(), 0, strings[3].length, 0)
            }
        }

        holder.mImageView.setImageResource(MiscData.getIconForCourse(currentItem.course))
        holder.mGroup.setText(strings[0], TextView.BufferType.SPANNABLE)
        holder.mTime.setText(strings[1], TextView.BufferType.SPANNABLE)
        holder.mCourse.setText(strings[2], TextView.BufferType.SPANNABLE)
        holder.mRoom.setText(strings[3], TextView.BufferType.SPANNABLE)
        holder.mAdditional.text = currentItem.additional

        if (currentItem.date.isNotEmpty() && currentItem.date.substring(0, 3) == "psa") {
            psa = true
            holder.mDate.visibility = View.GONE
            holder.mDate.text = currentItem.date
            holder.mTime.text = " "
            holder.mImageView.setImageResource(R.drawable.ic_idea)
            holder.mCard.setCardBackgroundColor(ContextCompat.getColor(holder.mImageView.context, R.color.colorAccent))
        } else {
            holder.mDate.visibility = View.VISIBLE
            holder.mDate.text = currentItem.date
        }

        if (!psa) {
            colourString = getColourString(holder.mCourse.text.toString())
            val colourPrefsInt = if (colourString.isNotEmpty()) {
                prefs.getString("card$colourString", "") ?: ""
            } else {
                ""
            }
            colour = MiscData.getColourForString(colourPrefsInt)
            if (colour != 0) {
                holder.mCard.setCardBackgroundColor(ContextCompat.getColor(holder.mCourse.context, colour))
            } else {
                holder.mCard.setCardBackgroundColor(ContextCompat.getColor(holder.mCourse.context, R.color.colorBackgroundLight))
            }
        }

        val textColor = ContextCompat.getColor(holder.mImageView.context, if (psa) {
            R.color.colorBackground
        } else {
            when (colour) {
                R.color.bgPureWhite -> R.color.colorTextDark
                R.color.bgPureBlack -> R.color.colorTextLight
                else -> R.color.colorText
            }
        })
        holder.mImageView.setColorFilter(textColor)
        holder.mGroup.setTextColor(textColor)
        holder.mDate.setTextColor(textColor)
        holder.mTime.setTextColor(textColor)
        holder.mCourse.setTextColor(textColor)
        holder.mRoom.setTextColor(textColor)
        holder.mAdditional.setTextColor(textColor)
    }

    override fun getItemCount(): Int = mSubst.size

    fun setSubst(subst: List<Subst>) {
        mSubst = subst
        notifyDataSetChanged()
    }

    private fun getColourString(course: String): String {
        return try {
            colorCheck = course.toLowerCase(Locale.ROOT).substring(0, 3)
            when (colorCheck) {
                "deu", "dep", "daz", "fda" -> "German"
                "mat", "map" -> "Maths"
                "eng", "enp", "ena" -> "English"
                "spo", "spp", "spth" -> "PhysEd"
                "pol", "pop" -> "Politics"
                "dar", "dap" -> "Theatre"
                "phy", "php" -> "Physics"
                "bio", "bip", "nw1", "nw2", "nw3", "nw4" -> "Biology"
                "che", "chp" -> "Chemistry"
                "phi", "psp" -> "Philosophy"
                "laa", "laf", "lat" -> "Latin"
                "spa", "spf" -> "Spanish"
                "fra", "frf", "frz" -> "French"
                "inf" -> "Compsci"
                "ges" -> "History"
                "rel" -> "Religion"
                "geg" -> "Geography"
                "kun" -> "Arts"
                "mus" -> "Music"
                "tue" -> "Turkish"
                "chi" -> "Chinese"
                "gll" -> "GLL"
                "wat" -> "WAT"
                "för" -> "Forder"
                "met", "wpb" -> "WP"
                else -> ""
            }
        } catch (e: StringIndexOutOfBoundsException) {
            try {
                colorCheck = course.toLowerCase(Locale.ROOT).substring(0, 2)
                when (colorCheck) {
                    "nw" -> "Biology"
                    "wp" -> "WP"
                    else -> ""
                }
            } catch (e2: StringIndexOutOfBoundsException) {
                ""
            }
        }
    }

}